### PR TITLE
1.4.3.2.1 Feature branch End Point

### PR DESCRIPTION
**Pull request by jasmeetkohli**
# Summary

API endpoint that retrieves FeatureBranches within a TimeFrame
syntax :` host/api/feature_branches?filter=none&start_time=123&end_time=1234`

# Requirements
There must be a Controller which defines methods for endpoints.

There must be a service and its implementation that takes care of the business logic for FeatureBranch.

There must be appropriate unit test cases for the controller and the service implementation.

# Changes

Created a controller `FeatureBranchController` which defines appropriate methods for endpoints.

Created a Service `FeatureBranchService` and `FeatureBranchServiceImpl` that takes care of the business logic.

Created two test files `FeatureBranchControllerTest` and `FeatureBranchServiceImplTest` that define test cases to verify the correct behaviour of endpoints and service implementation methods respectively.

### CODE REVIEW

File: api/src/main/java/com/capitalone/dashboard/rest/FeatureBranchController.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+public class FeatureBranchController {
+    
+    private final FeatureBranchService featureBranchService;
+
+    @Autowired
+    public FeatureBranchController(FeatureBranchService featureBranchService) {
+        this.featureBranchService = featureBranchService;
+    }
+
+    @RequestMapping(value = "/featurebranchbytimeframe/{timestamp1}/{timestamp2}", method = GET, produces = APPLICATION_JSON_VALUE)
```
Comment (mrrusof) [inline api/src/main/java/com/capitalone/dashboard/rest/FeatureBranchController.java]: Given that we are exposing feature branches and then filtering by timeframe, would it make sense to have the following syntax?

```
/feature_branches?start_timestamp=1234&end_timestamp=9876
```

**Review by mrrusof — CHANGES_REQUESTED** [review]

**Comment by jaloare** [comment]
Looks good .. I hope you can validate your code with SonarQube [it is ok if not]

File: api/src/main/java/com/capitalone/dashboard/service/FeatureBranchService.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+public interface FeatureBranchService {
+
+    List<FeatureBranch> getFeatureBranchByTimeFrame(long timestamp1, long timestamp2);
+    
+    List<FeatureBranch> getFeatureBranchByFirstCommitTimeFrame(long timestamp1, long timestamp2);
+    
+    List<FeatureBranch> getFeatureBranchByDeployTimeFrame(long timestamp1, long timestamp2);
+    
+    void save(FeatureBranch featureBranch);
```
Comment (mrrusof) [inline api/src/main/java/com/capitalone/dashboard/service/FeatureBranchService.java]: Who is using this?

File: api/src/main/java/com/capitalone/dashboard/service/FeatureBranchService.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+    Iterable<FeatureBranch> findAll();
```
Comment (mrrusof) [inline api/src/main/java/com/capitalone/dashboard/service/FeatureBranchService.java]: Who uses this?

File: api/src/main/java/com/capitalone/dashboard/service/FeatureBranchService.java (hunk 2 above)
Comment (mrrusof) [inline api/src/main/java/com/capitalone/dashboard/service/FeatureBranchService.java]: Who uses this?

**Review by mrrusof — CHANGES_REQUESTED** [review]

File: api/src/test/java/com/capitalone/dashboard/service/FeatureBranchServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+    @Test
+    public void validateSave() {
+
+        List<FeatureBranch> featureBranchList = getFeatureBranchList();
+
+        when(featureBranchRepository.findAll()).thenReturn(featureBranchList);
+
+        assertEquals("Save function validation for the FeatureBranchRepository has failed",
+                featureBranchService.findAll().size(),3);
```
Comment (mrrusof) [inline api/src/test/java/com/capitalone/dashboard/service/FeatureBranchServiceImplTest.java]: Did you mean? https://stackoverflow.com/questions/11598977/get-size-of-an-iterable-in-java#11599011

**Review by mrrusof — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/capitalone/dashboard/rest/FeatureBranchController.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            case "deploy" :
+                featureBranchList = featureBranchService.getFeatureBranchByDeployTimeFrame(startTime, endTime);
+                break;
+
+            case "none" :
+                featureBranchList = featureBranchService.getFeatureBranchByTimeFrame(startTime, endTime);
+                break;
+
+            default :
+                featureBranchList = new ArrayList();
```
Comment (mrrusof) [inline api/src/main/java/com/capitalone/dashboard/rest/FeatureBranchController.java]: Why not return HTTP 400 in this case?

**Review by mrrusof — CHANGES_REQUESTED** [review]

**Review by mrrusof — APPROVED** [review]
It delivers the requested functionality.

**Pull request merged**